enterprise/server/api/api_server.go

```diff
@@ -166,32 +164,50 @@ func (s *APIServer) GetAction(ctx context.Context, req *apipb.GetActionRequest)
 	if req.GetSelector().GetInvocationId() == "" {
 		return nil, status.InvalidArgumentErrorf("ActionSelector must contain a valid invocation_id")
 	}
+	iid := req.GetSelector().GetInvocationId()
+	rsp := &apipb.GetActionResponse{
+		Action: make([]*apipb.Action, 0),
+	}
+
+	// This limit is a safeguard against buffering too many results in memory.
+	// We expect to hit this rarely (if ever) in production usage.
+	const limit = 100_000
+	if *enableCache && s.env.GetMetricsCollector() != nil {
+		serializedResults, err := s.env.GetMetricsCollector().ListRange(ctx, api_common.ActionsKey(iid), 0, limit-1)
+		if err == nil {
+			a := &apipb.Action{}
+			for _, serializedResult := range serializedResults {
```

Comment:
Do we need to filter this down to just actions that match the selector?

Reply:
Done.